src/libopensc/card-openpgp.c

```diff
@@ -956,6 +960,83 @@ pgp_get_card_features(sc_card_t *card)
 			}
 		}
 
+#ifdef ENABLE_OPENSSL
+		if (priv->ext_caps & EXT_CAP_KDF_DO) {
+			pgp_blob_t *kdf_blob = NULL;
+			if ((pgp_get_blob(card, priv->mf, DO_KDF, &kdf_blob) == SC_SUCCESS) && (kdf_blob->len >= 3)) {
+				uint8_t *data = kdf_blob->data;
+				priv->pin_kdf_info = NULL;
+				/* KDF_ITERSALTED_S2K */
+				size_t i = 0;
+				pgp_pin_kdf_info *pin_kdf_info = (pgp_pin_kdf_info *)calloc(1, sizeof(pgp_pin_kdf_info));
+				if (!priv) {
+					free(priv);
+					pgp_free_blob(kdf_blob);
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+				}
+				int valid = 0;
+				if ((data[i] == 0x81) && (data[i + 1] == 0x01) && (data[i + 2] == 0x3)) {
+					valid = 1;
+					i += 3;
+				}
+
+				if (valid && (i + 3 <= kdf_blob->len)) {
+					valid = 0;
+					if (data[i] == 0x82 && data[i + 1] == 1) {
+						if (data[i + 2] == 0x08) {
+							pin_kdf_info->hash_algo = "SHA256";
+							valid = 1;
+						}
+						if (data[i + 2] == 0x0A) {
+							pin_kdf_info->hash_algo = "SHA512";
+							valid = 1;
+						}
+					}
+					i += 3;
+				}
+
+				if (valid && (i + 6 <= kdf_blob->len)) {
+					valid = 0;
+					if (data[i] == 0x83 && data[i + 1] == 0x04) {
+						pin_kdf_info->iterations = (uint32_t)bebytes2ulong(&data[i + 2]);
+						valid = 1;
+					}
+					i += 6;
+				}
```

Comment:
please use the parsing functions from asn1.h to avoid doing all the above by hand.